### PR TITLE
feat(rawdb&rpc): add block row consumption rpc

### PR DESCRIPTION
**Pull request by NazariiDenha**
## 1. Purpose or design rationale of this PR

We will call the "zkEVM capacity API" to decide if a transaction needs to be skipped or not. This API will be used during
block creation: make sure that the block does not cause "proof overflow",
block validation: make sure that blocks sent by peers fit into the zkEVM circuits.

When proposing chunks/batches, `rollup-relayer` will also need to make sure that a chunk fits into the zkEVM. Instead of calling this expensive API for each block again, we could store the "row acount" for each block during the above 2 steps in `l2geth`. Then, `rollup-relayer` can get this information through an l2geth RPC API.

Related discussion: https://scrollco.slack.com/archives/C02PKAQ2430/p1688486201888819
Asana tracker: https://app.asana.com/0/1202293017617135/1204864503012423/f

## 2. PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [ ] build: Changes that affect the build system or external dependencies (example scopes: yarn, eslint, typescript)
- [ ] ci: Changes to our CI configuration files and scripts (example scopes: vercel, github, cypress)
- [ ] docs: Documentation-only changes
- [x] feat: A new feature
- [ ] fix: A bug fix
- [ ] perf: A code change that improves performance
- [ ] refactor: A code change that doesn't fix a bug, or add a feature, or improves performance
- [ ] style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
- [ ] test: Adding missing tests or correcting existing tests

## 3. Deployment tag versioning

Has the version in `params/version.go` been updated?

- [ ] This PR doesn't involve a new deployment, git tag, docker image tag, and it doesn't affect traces
- [x] Yes

## 4. Breaking change label

Does this PR have the `breaking-change` label?

- [ ] This PR is not a breaking change
- [ ] Yes

### CODE REVIEW

File: internal/web3ext/web3ext.go [hunk 1]
```diff
@@ -874,7 +874,13 @@ web3._extend({
 			name: 'getFirstQueueIndexNotInL2Block',
 			call: 'scroll_getFirstQueueIndexNotInL2Block',
 			params: 1
-		})
+		}),
+		new web3._extend.Method({
+			name: 'getBlockByHash',
+			call: 'scroll_getBlockByHash',
```
Comment (HAOYUatHZ) [inline internal/web3ext/web3ext.go:880]: we already have `eth_getBlockByHash` why do we want `scroll_getBlockByHash`?   
or is it a mistake and should be `scroll_getBlockRowConsumptionByHash`

**Review by 0xmountaintop — DISMISSED** [review]

File: eth/api.go [hunk 2]
```diff
@@ -671,3 +671,30 @@ func (api *ScrollAPI) GetLatestRelayedQueueIndex(ctx context.Context) (queueInde
...
+// rpcMarshalBlock uses the generalized output filler, then adds the total difficulty field, which requires
+// a `ScrollAPI`.
+func (api *ScrollAPI) rpcMarshalBlock(ctx context.Context, b *types.Block, inclTx bool, fullTx bool) (map[string]interface{}, error) {
+	fields, err := ethapi.RPCMarshalBlock(b, inclTx, fullTx, api.eth.APIBackend.ChainConfig())
+	if err != nil {
+		return nil, err
+	}
+	if inclTx {
+		fields["totalDifficulty"] = (*hexutil.Big)(api.eth.APIBackend.GetTd(ctx, b.Hash()))
+	}
```
Comment (Thegaram) [inline eth/api.go]: Why do we need to add this here?

File: eth/api.go [hunk 3]
```diff
@@ -671,3 +671,30 @@ func (api *ScrollAPI) GetLatestRelayedQueueIndex(ctx context.Context) (queueInde
...
+	rc := rawdb.ReadBlockRowConsumption(api.eth.ChainDb(), b.Hash())
+	if rc != nil {
+		fields["rowConsumption"] = hexutil.Uint64(rc.Rows)
+	}
```
Comment (Thegaram) [inline eth/api.go]: If we add `else fields["rowConsumption"] = nil`, would that be serialized into JSON? If yes, that might be better, at least it's more explicit when this data is missing.

File: eth/api.go [hunk 4]
```diff
@@ -671,3 +671,30 @@ func (api *ScrollAPI) GetLatestRelayedQueueIndex(ctx context.Context) (queueInde
...
+	return fields, err
+}
+
+// GetBlockByHash returns the requested block. When fullTx is true all transactions in the block are returned in full
+// detail, otherwise only the transaction hash is returned.
+func (api *ScrollAPI) GetBlockByHash(ctx context.Context, hash common.Hash, fullTx bool) (map[string]interface{}, error) {
```
Comment (Thegaram) [inline eth/api.go:694]: Please also add to `ethclient` (see `GetBlockTraceByHash`) so that we can call this through the SDK.

**Review by Thegaram — COMMENTED** [review]
LGTM. Please also bump the version. How did you test this API?

File: eth/api.go (hunk 2 above)
Comment (NazariiDenha) [reply eth/api.go]: It comes from here https://github.com/scroll-tech/go-ethereum/blob/38a6a1e964e56d1cd8abb0f1bd4e3591385645e9/internal/ethapi/api.go#L1307

File: eth/api.go (hunk 3 above)
Comment (NazariiDenha) [reply eth/api.go]: Yes, added in new commit

File: eth/api.go (hunk 4 above)
Comment (NazariiDenha) [reply eth/api.go:694]: added in last commits

**Comment by NazariiDenha** [comment]
> LGTM. Please also bump the version. How did you test this API?

To test it locally added an RPC that writes `RowConsumption` to db for some block. And then read it through new `scroll_GetBlockByHash`

File: eth/api.go (hunk 2 above)
Comment (Thegaram) [reply eth/api.go]: If we only pass `true` we might as well remove the `if`.

Comment (NazariiDenha) [reply eth/api.go]: Ok, I see what you mean

**Review by Thegaram — APPROVED** [review]

**Review by 0xmountaintop — APPROVED** [review]

**Pull request merged**